### PR TITLE
Integrate async AVM, introduce user-defined callbacks

### LINKED ISSUE #1192: Integrate async AquaVM
AquaVM has changed it's interfaces: it's no longer calls host functions imperatively, but instead declaratively asks the node to run them. This allows the node to utilize AquaVM pool in efficiently: it's mostly CPU-bound now, and all the heavy host/service calls can be moved to a separate thread pool.

### PR DESCRIPTION
**Pull request by folex**
fixes #1192 

TODOs:
- [x] Cleanup data on actor expiration https://github.com/fluencelabs/fluence/issues/1216
- [x] Make actor.particle non-optional
- [ ] Create vault in actor/functions https://github.com/fluencelabs/fluence/issues/1217
- [ ] Ditch ConnectedClient and use Node as a client instead https://github.com/fluencelabs/fluence/issues/1218
- [x] ~Implement ParticleFunction for Fn~ (doesn't seem to be possible)
- [x] Tidy code in BuiltinsDeployer
- [x] Update to latest AVM and interpreter

### CODE REVIEW

File: aquamarine/src/deadline.rs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+use particle_protocol::Particle;
+use std::ops::Mul;
+
+#[derive(Debug, Clone)]
+pub struct Deadline {
+    timestamp: u64,
```
Comment (justprosh) [inline aquamarine/src/deadline.rs:22]: maybe we should leave a comment that timestamp in secs and ttl in millis

File: Cargo.toml [hunk 2]
```diff
@@ -30,11 +30,43 @@ members = [
...
+#
+#
+#[patch.'https://github.com/fluencelabs/aquavm']
+#marine-macro-impl = { version = "0.6.12", git = "https://github.com/fluencelabs/marine-rs-sdk", branch = "master" }
+#marine-macro = { git = "https://github.com/fluencelabs/marine-rs-sdk", branch = "master" }
+#marine-rs-sdk = { git = "https://github.com/fluencelabs/marine-rs-sdk", branch = "master" }
+#marine-rs-sdk-main = { git = "https://github.com/fluencelabs/marine-rs-sdk", branch = "master" }
+
+
+[patch.crates-io]
```
Comment (justprosh) [inline Cargo.toml:57]: Don't forget to remove

File: aquamarine/src/actor.rs [hunk 3]
```diff
@@ -117,39 +124,57 @@ where
...
+
+        if particle.is_none() && calls.is_empty() {
+            // Nothing to execute, return vm
+            return ActorPoll::Vm(vm);
         }
+
+        // SAFETY: At least one particle was ingested or calls/mailbox would be empty.
+        let particle = particle.or_else(|| self.particle.clone()).unwrap();
+        let waker = cx.waker().clone();
+        // TODO: add timeout for execution
```
Comment (justprosh) [inline aquamarine/src/actor.rs:148]: Let's add issue

File: aquamarine/src/plumber.rs [hunk 4]
```diff
@@ -112,16 +117,33 @@ impl<RT: AquaRuntime> Plumber<RT> {
             mailbox_size += actor.mailbox_size();
         }
 
+        // Remove expired actors
+        let now = now_ms();
+        self.actors.retain(|particle_id, actor| {
+            // if actor hasn't yet expired or is still executing, keep it
+            // TODO: if actor is expired, cancel execution and return VM back to pool
```
Comment (justprosh) [inline aquamarine/src/plumber.rs:124]: Ditto

File: crates/created-swarm/src/swarm.rs [hunk 5]
```diff
@@ -295,133 +283,218 @@ pub fn aqua_vm_config(
...
-
-    let host_closures = Node::host_closures(
-        connectivity,
-        vec![listen_on],
-        services_config,
-        script_storage_api,
-    );
-
-    (vm_config, host_closures.descriptor())
+    // let services_config = ServicesConfig::new(
```
Comment (justprosh) [inline crates/created-swarm/src/swarm.rs:286]: dead code?

File: crates/created-swarm/src/swarm.rs [hunk 6]
```diff
@@ -295,133 +283,218 @@ pub fn aqua_vm_config(
...
-            peer_id,
-        };
+    resolved.prometheus_config.prometheus_enabled = false;
 
-        let pool: &ConnectionPoolApi = connectivity.as_ref();
-        ScriptStorageBackend::new(pool.clone(), particle_failures_in, script_storage_config)
-    };
+    resolved.node_config.allow_local_addresses = true;
+    resolved.node_config.particle_processing_timeout = Duration::from_secs(45);
+    // resolved.node_config.listen_config.listen_multiaddrs = vec![.clone()];
```
Comment (justprosh) [inline crates/created-swarm/src/swarm.rs:356]: ditto

File: crates/created-swarm/src/swarm.rs [hunk 7]
```diff
@@ -295,133 +283,218 @@ pub fn aqua_vm_config(
...
+        // to_peer_id(resolved.builtins_key_pair.clone().into()),
     );
+    let mut node = Node::new(resolved, vm_config, "some version").expect("create node");
+    node.listen(vec![config.listen_on.clone()]).expect("listen");
 
-    node.listen(vec![listen_on]).expect("listen");
+    (node.local_peer_id, node, management_kp, config)
 
-    (peer_id, node, management_kp, config)
+    // TODO: this requires ability to convert public keys to PeerId
```
Comment (justprosh) [inline crates/created-swarm/src/swarm.rs:381]: New fluence-keypair crate can do this

File: crates/created-swarm/src/swarm.rs [hunk 8]
```diff
@@ -295,133 +283,218 @@ pub fn aqua_vm_config(
...
+    // if let Some(trust) = config.trust {
+    //     node_config.root_weights = trust.root_weights.into_iter().map(|(pk, weight)| {
+    //         pk.
+    //     }).collect();
+    // }
 }
 
-pub fn create_swarm(config: SwarmConfig) -> (PeerId, Box<Node<AVM>>, Keypair, SwarmConfig) {
+// pub fn create_swarm_with_runtime<RT: AquaRuntime>(
```
Comment (justprosh) [inline crates/created-swarm/src/swarm.rs:389]: dead code

File: crates/particle-args/Cargo.toml [hunk 9]
```diff
@@ -9,11 +9,12 @@ ivalue-utils = { path = "../ivalue-utils" }
 json-utils = { path = "../json-utils" }
 control-macro = { path = "../control-macro" }
 
-marine-runtime = "0.7.0"
-avm-server = "0.10.0"
-fluence-app-service = "0.9.0"
+marine-runtime = { git = "https://github.com/fluencelabs/marine", branch = "master" }
+avm-server = "0.11.0"
+#fluence-app-service = "0.10.1"
```
Comment (justprosh) [inline crates/particle-args/Cargo.toml:14]: dead code

File: crates/server-config/src/resolved_config.rs [hunk 10]
```diff
@@ -151,11 +151,19 @@ impl ResolvedConfig {
...
-        SocketAddr::new(self.listen_ip, self.prometheus_port)
+        SocketAddr::new(
+            self.listen_config.listen_ip,
+            self.prometheus_config.prometheus_port,
+        )
     }
 
     pub fn listen_config(&self) -> ListenConfig {
-        ListenConfig::new(self.listen_ip, self.tcp_port, self.websocket_port)
+        // TODO: only 1 ListenConfig must survive
```
Comment (justprosh) [inline crates/server-config/src/resolved_config.rs:161]: let's file an issue

File: particle-execution/Cargo.toml [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+particle-protocol = { path = "../particle-protocol" }
+
+particle-args = { path = "../crates/particle-args" }
+fluence-libp2p = { path = "../crates/libp2p" }
+fs-utils = { path = "../crates/fs-utils" }
+json-utils = { path = "../crates/json-utils" }
+
+thiserror = "1.0.24"
+futures = "0.3.13"
+serde_json = "1.0.64"
```
Comment (justprosh) [inline particle-execution/Cargo.toml:16]: missing newline

File: particle-execution/src/particle_params.rs [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+use particle_protocol::Particle;
+
+/// Lightweight, static version of the [[Particle]] structure
+/// It exists to avoid cloning [[Particle::data]] when possible
+#[derive(Debug, Clone)]
+pub struct ParticleParams {
+    pub id: String,
+    pub init_peer_id: PeerId,
+    pub timestamp: u64,
+    pub ttl: u32,
```
Comment (justprosh) [inline particle-execution/src/particle_params.rs:27]: secs or millis? Should be documented i think

File: particle-node/benches/aquamarine_bench.rs [hunk 13]
```diff
@@ -1,96 +1,95 @@
...
+// use criterion::Criterion;
+// use criterion::{criterion_group, criterion_main};
+// use fluence_libp2p::RandomPeerId;
+// use local_vm::{make_call_service_closure, make_vm};
+// use maplit::hashmap;
+// use parking_lot::Mutex;
+// use serde_json::{json, Value as JValue};
+// use std::collections::HashMap;
+// use std::sync::Arc;
+//
```
Comment (justprosh) [inline particle-node/benches/aquamarine_bench.rs:26]: dead code

File: particle-node/benches/bench_network_models.rs [hunk 14]
```diff
@@ -1,457 +1,455 @@
...
+//  * Unless required by applicable law or agreed to in writing, software
+//  * distributed under the License is distributed on an "AS IS" BASIS,
+//  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  * See the License for the specific language governing permissions and
+//  * limitations under the License.
+//  */
+//
+// #![allow(dead_code)]
+//
+// use std::mem;
```
Comment (justprosh) [inline particle-node/benches/bench_network_models.rs:19]: ditto

File: particle-node/benches/kademlia_test.rs [hunk 15]
```diff
@@ -1,136 +1,128 @@
...
+//     unused_mut,
+//     unused_variables,
+//     unused_unsafe,
+//     unreachable_patterns
+// )]
+//
+// mod bench_network_models;
+//
+// use bench_network_models::*;
+// use connected_client::ConnectedClient;
```
Comment (justprosh) [inline particle-node/benches/kademlia_test.rs:30]: ditto

File: particle-node/src/connectivity.rs [hunk 16]
```diff
@@ -0,0 +1,218 @@
...
+                    let addresses = iter(addresses.collect::<Vec<_>>());
+                    return Some(addresses);
+                }
+                None
+            })
+        }
+        .flatten();
+
+        // TODO: take from config
+        let max = Duration::from_secs(60);
```
Comment (justprosh) [inline particle-node/src/connectivity.rs:180]: Maybe define as constants for now

File: particle-node/src/effectors.rs [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+                // resolve contact
+                if let Some(contact) = connectivity.resolve_contact(target, &particle.id).await {
+                    // forward particle
+                    connectivity.send(contact, particle).await;
+                }
+            }
+        })
+        .await;
+
+        // let crs = effects.call_requests;
```
Comment (justprosh) [inline particle-node/src/effectors.rs:60]: dead code

File: particle-node/src/node.rs [hunk 18]
```diff
@@ -45,70 +32,104 @@ use libp2p::{
...
+
     registry: Option<Registry>,
     metrics_listen_addr: SocketAddr,
-    bootstrap_nodes: Vec<Multiaddr>,
-    particle_failures: Outlet<String>,
-    script_storage: ScriptStorageBackend,
-    pub startup_management_peer_id: PeerId,
+
+    pub local_peer_id: PeerId,
+    // TODO: remove?
```
Comment (justprosh) [inline particle-node/src/node.rs:79]: why?

File: particle-node/src/node.rs [hunk 19]
```diff
@@ -262,11 +334,18 @@ impl<RT: AquaRuntime> Node<RT> {
...
             log::info!("Stopping node");
             script_storage.cancel().await;
-            network.cancel().await;
+            dispatcher.cancel().await;
+            connectivity.cancel().await;
             pool.cancel().await;
         });
 
-        exit_outlet
+        // TODO: make `deploy_builtin_services` async and add to the "select! loop" above
```
Comment (justprosh) [inline particle-node/src/node.rs:342]: let's file an issue

File: particle-node/benches/network_api_bench.rs [hunk 20]
```diff
@@ -1,316 +1,300 @@
...
+//  * Unless required by applicable law or agreed to in writing, software
+//  * distributed under the License is distributed on an "AS IS" BASIS,
+//  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  * See the License for the specific language governing permissions and
+//  * limitations under the License.
+//  */
+//
+// mod bench_network_models;
+//
+// use bench_network_models::*;
```
Comment (justprosh) [inline particle-node/benches/network_api_bench.rs:19]: dead code

File: particle-node/src/tasks.rs [hunk 21]
```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 2020 Fluence Labs Limited
```
Comment (justprosh) [inline particle-node/src/tasks.rs:2]:
```suggestion
 * Copyright 2021 Fluence Labs Limited
```

File: particle-node/tests/builtins/tests/src/aqua_dht/pubsub.rs [hunk 22]
```diff
@@ -34,10 +37,15 @@ fn find_subscribers() {
         .wrap_err("connect client")
         .unwrap();
 
-    log::info!("gonna send particle");
-    // initTopicAndSubscribe(node_id: PeerId, topic: string, value: string, relay_id: ?PeerId, service_id: ?string):
+    // log::info!("client: {} => {}", client.peer_id, client.node);
```
Comment (justprosh) [inline particle-node/tests/builtins/tests/src/aqua_dht/pubsub.rs:40]: debug logs?

File: crates/builtins-deployer/src/builtins_deployer.rs [hunk 23]
```diff
@@ -315,9 +198,14 @@ impl BuiltinsDeployer {
...
 
-        assert_ok(result, "create_service call failed")
+        assert_ok(result, "create_service call failed")?;
+
+        log::info!("service was created!");
+
+        Ok(())
     }
 
+    // TODO: right now, if AIR in on_start.air is invalid, everything just hangs
```
Comment (justprosh) [inline crates/builtins-deployer/src/builtins_deployer.rs:208]: let's file an issue

**Review by justprosh — APPROVED** [review]

File: aquamarine/src/actor.rs (hunk 3 above)
Comment (folex) [reply aquamarine/src/actor.rs:148]: https://github.com/fluencelabs/marine/issues/131

File: aquamarine/src/plumber.rs (hunk 4 above)
Comment (folex) [reply aquamarine/src/plumber.rs:124]: https://github.com/fluencelabs/fluence/issues/1212

File: crates/created-swarm/src/swarm.rs (hunk 7 above)
Comment (folex) [reply crates/created-swarm/src/swarm.rs:370]: It's too much work to update fluence-keypair & trust-graph crates. https://github.com/fluencelabs/fluence/issues/1213

File: particle-node/src/connectivity.rs (hunk 16 above)
Comment (folex) [reply particle-node/src/connectivity.rs:180]: what's the difference?

File: particle-node/src/node.rs (hunk 19 above)
Comment (folex) [reply particle-node/src/node.rs:342]: I don't remember why I wanted that :(

File: crates/builtins-deployer/src/builtins_deployer.rs (hunk 23 above)
Comment (folex) [reply crates/builtins-deployer/src/builtins_deployer.rs:208]: https://github.com/fluencelabs/fluence/issues/1214

File: crates/builtins-deployer/src/builtins_deployer.rs [hunk 24]
```diff
@@ -206,6 +206,7 @@ impl BuiltinsDeployer {
     }
 
     // TODO: right now, if AIR in on_start.air is invalid, everything just hangs
+    //       https://github.com/fluencelabs/fluence/issues/1214
```
Comment (justprosh) [inline crates/builtins-deployer/src/builtins_deployer.rs:209]:
```suggestion
    // https://github.com/fluencelabs/fluence/issues/1214
```

File: crates/created-swarm/Cargo.toml [hunk 25]
```diff
@@ -20,6 +20,7 @@ air-interpreter-fs = { path = "../air-interpreter-fs" }
 builtins-deployer = { path = "../builtins-deployer" }
 toy-vms = { path = "../toy-vms" }
 
+
```
Comment (justprosh) [inline crates/created-swarm/Cargo.toml:23]:
```suggestion
```

File: crates/server-config/Cargo.toml [hunk 26]
```diff
@@ -12,6 +12,7 @@ fluence-libp2p = { path = "../libp2p" }
 air-interpreter-fs = { path = "../air-interpreter-fs" }
 
 trust-graph = "0.2.7"
+
```
Comment (justprosh) [inline crates/server-config/Cargo.toml:15]: is it intended?

File: particle-node/tests/builtins/Cargo.toml [hunk 27]
```diff
@@ -12,6 +12,7 @@ serde = "1.0.130"
 log = "0.4.11"
 libp2p = { version = "0.36.2", package = "fluence-fork-libp2p", features = ["tcp-tokio"] }
 
+
```
Comment (justprosh) [inline particle-node/tests/builtins/Cargo.toml:15]:
```suggestion
```

**Review by justprosh — APPROVED** [review]

File: Cargo.toml (hunk 2 above)
Comment (folex) [reply Cargo.toml:43]: I'll keep trust-graph and fluence-identity patches for now.

File: particle-node/benches/network_api_bench.rs (hunk 20 above)
Comment (folex) [reply particle-node/benches/network_api_bench.rs:19]: https://github.com/fluencelabs/fluence/issues/1219

**Pull request merged**